### PR TITLE
Introduce `AuthorityVerifier` to handle blockchain signature verification

### PR DESCRIPTION
**Pull request by prince-chrismc**
## PR Checklist

- [x] I've built the code `cargo build`. For major changes, `cargo build --workspace --release` is recommended.
- [x] I've run the automated unit tests `cargo test`. This executes our automated unit tests.
- [x] I've not broken any existing tests or functionality. In addition to `cargo test`, I've run `cargo clippy`
- [x] I've not introduced any new known security vulnerabilities. I've run `cargo audit`

- [x] I've included a brief description and a good title of the proposed changes and how to test them.
- [x] I've read the [contributing guidelines](https://github.com/pyrsia/.github/blob/main/contributing.md).
- [x] I've associated an [issue](https://github.com/pyrsia/pyrsia/issues) with this Pull Request. 
- [x] I've checked that the code is up-to-date with the `pyrsia/main` branch.
- [x] I've read ["What is a Good PR?"](https://github.com/pyrsia/pyrsia/blob/main/docs/good_pr.md)
- [x] I've assigned this Pull Request to "pyrsia/collaborators"

## Description

Towards #322 

along with some support structs and tests.
everything is grouped into an `identities` module

The end goal is the `KeyBox` https://github.com/prince-chrismc/pyrsia-sandbox/blob/a9333aced792e1017db1f424eb1b4b77c7803aa4/src/block_chain/crypto.rs#L132 for aleph
which is "the guy who knows all the signatures and how to verify them" this is a helper for that.

### CODE REVIEW

File: src/blockchain/src/identities/authority_verifier.rs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        trace!(
+            "Recording new authority {:?} with {:?}",
+            node_ix,
+            public_key
+        );
+        self.authorities.insert(node_ix, public_key);
+    }
+    /// Verifies whether the message is correctly signed with the signature assumed to be made by a
+    /// node of the given index.
+    pub fn verify(&self, msg: &[u8], sgn: &Signature, index: NodeIndex) -> bool {
```
Comment (chb0github) [inline src/blockchain/src/identities/authority_verifier.rs:45]: Hmm, maybe it would be better to return `Result<bool,Error>`

**Review by chb0github — APPROVED** [review]
Just a suggestion

File: src/blockchain/src/identities/authority_verifier.rs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+use aleph_bft::{NodeCount, NodeIndex};
+use libp2p::core::identity::ed25519::PublicKey;
+use log::{trace, warn};
+use std::collections::HashMap;
+
+use super::signature::{MultiSignature, Signature};
+
+#[derive(Clone, Default)]
+pub struct AuthorityVerifier {
```
[the remark this answers is not in the document]
Comment (prince-chrismc) [reply src/blockchain/src/identities/authority_verifier.rs:25]: Yes you need to know of all of the verifiers for now. We can expand that later post MVP 👍

File: src/blockchain/src/identities/authority_verifier.rs (hunk 1 above)
Comment (prince-chrismc) [reply src/blockchain/src/identities/authority_verifier.rs:45]: The next PR will be `KeyBox` (as mentioned in the description) that is a trait for aleph.

This needs to be a `bool` for that API requirement.

https://github.com/prince-chrismc/pyrsia-sandbox/blob/a9333aced792e1017db1f424eb1b4b77c7803aa4/src/block_chain/crypto.rs#L151

Down the road we can investigate doing more 👍

**Review by efrisch — APPROVED** [review]
👍

File: src/blockchain/src/identities/authority_verifier.rs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+            }
+        }
+    }
+
+    pub fn node_count(&self) -> NodeCount {
+        self.authorities.len().into()
+    }
+
+    fn threshold(&self) -> usize {
+        2 * self.node_count().0 / 3 + 1
```
[the remark this answers is not in the document]
Comment (prince-chrismc) [reply src/blockchain/src/identities/authority_verifier.rs:61]: image


🤣 

This was copy paste https://github.com/Cardinal-Cryptography/aleph-node/blob/2192ed2c9a960c90fd86c17b121a60a3e59c6046/finality-aleph/src/crypto.rs#L100-L102

**Pull request merged**